Modules/_json.c

```diff
@@ -1643,19 +1684,30 @@ encoder_listencode_list(PyEncoderObject *s, _PyUnicodeWriter *writer,
 
     if (_PyUnicodeWriter_WriteChar(writer, '['))
         goto bail;
+
+    PyObject* separator = s->item_separator; // borrowed reference
+    PyObject* newline_indent = 0;
     if (s->indent != Py_None) {
-        /* TODO: DOES NOT RUN */
-        indent_level += 1;
-        /*
-            newline_indent = '\n' + (' ' * (_indent * _current_indent_level))
-            separator = _item_separator + newline_indent
-            buf += newline_indent
-        */
+        indent_level++;
+        newline_indent = _create_newline_indent(s->indent, indent_level);
+        if (newline_indent == 0) {
+            goto bail;
+        }
+
+        if (_PyUnicodeWriter_WriteStr(writer, newline_indent) < 0) {
+            goto bail;
+        }
+
+        separator = PyUnicode_Concat(separator, newline_indent); // non-borrowed reference
```

Comment:
I don't know that "non-borrowed reference" means here but it looks like we're leaking a reference to the previous contents of `separator`.

Reply:
It is okay in normal case. `separator` is a borrowed reference when `s->indent == Py_None` and a non-borrowed reference when `s->indent != Py_None`. It is balanced out in normal flow. But if an error happens and we go to the `bail` label, `separator` can be leaked. It should be decrefed if needed.